Forward to Logsearch if bosh link provided

Infra-Red opened the pull request:
Hi!
That would be great if we could configure rsyslog forwarding to Logsearch with syslog-release addon if BOSH link is provided in manifest. With this patch first Logsearch ingestor instance will be configured as syslog server and rest will be added to list of fallback servers.

Infra-Red commented:
@dpb587-pivotal Thanks for your comments! I updated PR to use generic `syslog` link, will also update Logsearch to provide `syslog` link and all necessary properties.

dpb587-pivotal commented:
Thanks for the quick update! Thinking about this more, `syslog_storer` is essentially doing the same thing. I wonder if we should just clean up the way the release currently provides that link and then make the forwarder slightly more generic naming-wise.

Also, at some point we'll want to move these jobs to use job-level properties (to avoid the extra `syslog.` prefix in the property names) to have a more concise `link('syslog').p('port')`-like syntax. If we're changing the link, might be a good time to do that.

Will give this a brief thought...

Infra-Red commented:
Hi @dpb587-pivotal ! I've updated `syslog_storer` job to unify name of provided link, now `syslog_forwarder` job will use `syslog` link if it's added to deployment manifest. I've also changed property names used for this link, but wanted to clarify did you mean all property names used by release to be changed or only properties used with `syslog` link.

dpb587-pivotal commented:
Nice work - this looks like what I was imagining. It's a bit of a breaking change for storer, so will require a bit more discussion. For the properties currently present, I like the links property API.

Optimally, we'll want to change the properties of forwarder as well, but I think it makes sense to wait on that change since it's used by others and will require some advanced planning and discussion. The changes can also happen independently of this.

I'll chat with @cppforlife about this soon and update/merge when I know more. Thanks again!

cppforlife commented on jobs/syslog_forwarder/templates/rsyslog.conf.erb:
```diff
@@ -9,10 +9,21 @@ if_p('syslog.address', 'syslog.port', 'syslog.transport') do |address, port, tra
+
+fallback_servers = nil
+respond_to?(:if_link) && if_link('syslog') do |syslog_link|
+  fallback_servers = syslog_link.instances.reject{ |instance| instance.index == 0 }.map { |instance|
```
`index == 0` is same as `instances[0]`. im assuming above is trying to eliminate address pulled out on line 13?

Infra-Red replied:
@cppforlife Yes, that's right.

cppforlife replied:
@Infra-Red oops i mean index == 0 is NOT same as instances[0].

Infra-Red replied:
@cppforlife Then I think it would be better to add a check for bootstrap instance and put all the rest in list of fallback servers. WDYT?

tylerschultz replied:
@Infra-Red There are scenarios where there will be not be an instance with index 0. There will always be a bootstrap instance. If bootstrap works accomplishes what you need, then absolutely, that would be better.

cppforlife replied:
sgtm.

dpb587-pivotal replied:
Since there's not a dependence on the vm actually being the bootstrap, I think it'd be better to ignore an instance if `instance.address == syslog_address`. Avoids coupling to the bootstrap concept and avoids an initial loop to find `syslog_address`.

Infra-Red commented:
Updated according to @dpb587-pivotal comment.
/cc @cppforlife @tylerschultz

dpb587-pivotal commented on jobs/syslog_forwarder/templates/rsyslog.conf.erb:
```diff
@@ -9,10 +9,21 @@ if_p('syslog.address', 'syslog.port', 'syslog.transport') do |address, port, tra
+  fallback_servers = syslog_link.instances.reject{ |instance| instance.address == syslog_address }.map { |instance|
+    { 'address' => instance.address,
+      'port' => link('syslog').p('port'),
+      'transport' => link('syslog').p('transport') }
```
Minor, instead of resolving the link every time, we can probably reuse `syslog_port` or `syslog_transport variables.

Infra-Red commented:
@dpb587-pivotal Any chance you can take a look at this PR anytime soon?

anEXPer commented:
We'd still be interested in taking this, but have made substantial changes to the way rsyslog config gets put in place.

We'd update the PR ourselves, but the team's been deallocated, for now. Still, we'd be interested in taking an updated version of this PR.

Sorry for the delay, it's been a turbulent time in syslog-release's life.

Infra-Red commented:
I've updated PR. @anEXPer could you please take a look at this? Thanks!

anEXPer commented:
I've taken a look, and it looks good to me. I'd like to add a test and get this merged, hopefully sometime next week. Thanks for updating the PR!
